File: app/src/main/java/com/lmyby/ankihelper/data/dict/BaseDatabaseHelper.java
package com.lmyby.ankihelper.data.dict;

import android.content.Context;
import android.content.SharedPreferences;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;

import com.lmyby.ankihelper.data.database.DatabaseContext;
import com.lmyby.ankihelper.util.StorageManager;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;

public class BaseDatabaseHelper extends SQLiteOpenHelper {

    private static final String TAG = "BaseDatabaseHelper";
    private final String databaseName;
    private final File databaseFile;
    private final StorageManager storageManager;
    private final Context context;

    public BaseDatabaseHelper(Context context, String databaseName, int version) {
        super(new DatabaseContext(context), databaseName, null, version);
        this.context = context;
        this.databaseName = databaseName;
        SharedPreferences preferences = context.getSharedPreferences("ankihelper_prefs", Context.MODE_PRIVATE);
        this.storageManager = new StorageManager(context, preferences);
        this.databaseFile = new File(storageManager.getDatabaseDir(), databaseName);

        Log.d(TAG, "Initializing dictionary database: " + databaseName);
        Log.d(TAG, "Dictionary database path: " + databaseFile.getAbsolutePath());

        // Force copy from assets if database doesn't exist or is empty
        copyDatabaseFromAssets(context);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        // Database should be copied from assets
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        // Handle upgrades if necessary
    }

    private void copyDatabaseFromAssets(Context context) {
        if (databaseFile.exists() && databaseFile.length() > 1000) {
            return; // Database already exists and has content
        }

        if (!databaseFile.getParentFile().exists()) {
            databaseFile.getParentFile().mkdirs();
        }

        try {
            InputStream input = context.getAssets().open("databases/" + databaseName);
            FileOutputStream output = new FileOutputStream(databaseFile);

            byte[] buffer = new byte[8192];
            int bytesRead;
            long totalBytes = 0;

            while ((bytesRead = input.read(buffer)) != -1) {
                output.write(buffer, 0, bytesRead);
                totalBytes += bytesRead;
            }

            input.close();
            output.close();

        } catch (IOException e) {
            Log.e(TAG, "Failed to copy " + databaseName + " from assets: " + e.getMessage());
            // Create minimal tables as fallback
            createFallbackDatabase();
        }
    }

    private void createFallbackDatabase() {
        SQLiteDatabase db = SQLiteDatabase.openOrCreateDatabase(databaseFile, null);
        db.execSQL("CREATE TABLE IF NOT EXISTS dict (hwd TEXT PRIMARY KEY, phrase TEXT, sense TEXT, phonetics TEXT, def_en TEXT, def_cn TEXT)");
        db.execSQL("CREATE TABLE IF NOT EXISTS hwds (rowid INTEGER PRIMARY KEY, hwd TEXT)");
        db.close();
    }

    @Override
    public SQLiteDatabase getReadableDatabase() {
        if (!databaseFile.exists() || databaseFile.length() < 1000) {
            // Database missing or corrupted, attempt rebuild
            if (!databaseFile.getParentFile().exists()) {
                databaseFile.getParentFile().mkdirs();
            }
            
            copyDatabaseFromAssets(context);
            
            if (!databaseFile.exists() || databaseFile.length() < 1000) {
                // Still no database, create fallback
                SQLiteDatabase db = SQLiteDatabase.openOrCreateDatabase(databaseFile, null);
                createFallbackDatabase();
                db.close();
            }
        }
        
        return super.getReadableDatabase();
    }
}

File: app/src/main/java/com/lmyby/ankiquicker/data/dict/BaseDatabaseHelper.kt
package com.lmyby.ankiquicker.data.dict

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.util.Log
import com.lmyby.ankiquicker.data.database.DatabaseContext
import com.lmyby.ankiquicker.util.StorageManager
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

/**
 * Converted to Kotlin as part of Phase 2 data model migration
 */
open class BaseDatabaseHelper(
    context: Context,
    private val databaseName: String,
    version: Int
) : SQLiteOpenHelper(DatabaseContext(context), databaseName, null, version) {

    private val context: Context = context
    private val storageManager: StorageManager
    private val databaseFile: File

    init {
        val preferences = context.getSharedPreferences("ankihelper_prefs", Context.MODE_PRIVATE)
        this.storageManager = StorageManager(context, preferences)
        this.databaseFile = File(storageManager.getDatabaseDir(), databaseName)

        Log.d(TAG, "Initializing dictionary database: $databaseName")
        Log.d(TAG, "Dictionary database path: ${databaseFile.absolutePath}")

        // Force copy from assets if database doesn't exist or is empty
        copyDatabaseFromAssets(context)
    }

    override fun onCreate(db: SQLiteDatabase) {
        // Database should be copied from assets
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        // Handle upgrades if necessary
    }

    private fun copyDatabaseFromAssets(context: Context) {
        if (databaseFile.exists() && databaseFile.length() > 1000) {
            return // Database already exists and has content
        }

        if (!databaseFile.parentFile!!.exists()) {
            databaseFile.parentFile!!.mkdirs()
        }

        try {
            context.assets.open("databases/$databaseName").use { input ->
                FileOutputStream(databaseFile).use { output ->
                    val buffer = ByteArray(8192)
                    var bytesRead: Int
                    var totalBytes: Long = 0

                    while (input.read(buffer).also { bytesRead = it } != -1) {
                        output.write(buffer, 0, bytesRead)
                        totalBytes += bytesRead.toLong()
                    }
                }
            }
        } catch (e: IOException) {
            Log.e(TAG, "Failed to copy $databaseName from assets: ${e.message}")
            // Create minimal tables as fallback
            createFallbackDatabase()
        }
    }

    private fun createFallbackDatabase() {
        val db = SQLiteDatabase.openOrCreateDatabase(databaseFile, null)
        db.execSQL("CREATE TABLE IF NOT EXISTS dict (hwd TEXT PRIMARY KEY, phrase TEXT, sense TEXT, phonetics TEXT, def_en TEXT, def_cn TEXT)")
        db.execSQL("CREATE TABLE IF NOT EXISTS hwds (rowid INTEGER PRIMARY KEY, hwd TEXT)")
        db.close()
    }

    override fun getReadableDatabase(): SQLiteDatabase {
        if (!databaseFile.exists() || databaseFile.length() < 1000) {
            // Database missing or corrupted, attempt rebuild
            if (!databaseFile.parentFile!!.exists()) {
                databaseFile.parentFile!!.mkdirs()
            }

            copyDatabaseFromAssets(context)

            if (!databaseFile.exists() || databaseFile.length() < 1000) {
                // Still no database, create fallback
                val db = SQLiteDatabase.openOrCreateDatabase(databaseFile, null)
                createFallbackDatabase()
                db.close()
            }
        }

        return super.getReadableDatabase()
    }

    companion object {
        private const val TAG = "BaseDatabaseHelper"
    }
}
